jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/authentication/FormAuthenticator.java

```diff
@@ -352,8 +362,55 @@ public AuthenticationState validateRequest(Request request, Response response, C
         // send the challenge
         if (LOG.isDebugEnabled())
             LOG.debug("challenge {}->{}", session.getId(), _formLoginPage);
-        Response.sendRedirect(request, response, callback, encodeURL(URIUtil.addPaths(request.getContext().getContextPath(), _formLoginPage), request), true);
-        return AuthenticationState.CHALLENGE;
+        return sendChallenge(request, response, callback);
+    }
+
+    private AuthenticationState dispatch(String path, Request request, Response response, Callback callback)
+    {
+        try
+        {
+            HttpURI.Mutable newUri = HttpURI.build(request.getHttpURI()).pathQuery(path);
+            return new AuthenticationState.ServeAs(newUri)
+            {
+                @Override
+                public Request wrap(Request request)
+                {
+                    org.eclipse.jetty.security.ServeAs serveAs = Request.as(request, org.eclipse.jetty.security.ServeAs.class);
+                    if (serveAs != null)
+                        return serveAs.serveAs(request, path);
+                    return super.wrap(request);
```

Comment:
Could this logic be in the base class?

Reply:
Probably not because the `super.wrap()` does not take a path.

And where we call the `ServeAs.wrap()` securityHandler we don't know which path it needs to go to anyway.